File: 15. 3Sum.cpp
/* 100 ms */
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> allSol;
        if(nums.size()<3) return allSol;
        sort(nums.begin(),nums.end());
        
        for(int i=0; i<nums.size()-2; i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int left = i+1;
            int right = nums.size()-1;
            while(left<right){
                if(nums[i]+nums[left]+nums[right] == 0){
                    vector<int> sol;
                    sol.push_back(nums[i]);
                    sol.push_back(nums[left]);
                    sol.push_back(nums[right]);
                    allSol.push_back(sol);
                    left++;
                    right--;
                    while(nums[left] == nums[left-1]) left++;
                    while(nums[right] == nums[right+1]) right--;
                }
                else if(nums[i]+nums[left]+nums[right] > 0) {
                    right--;
                }
                else if(nums[i]+nums[left]+nums[right] < 0) {
                    left++;
                }
            }
        }
        return allSol;        
    }
};

/**/
vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> ret;
        for(int i = 0;i<nums.size();i++){
            if(i>0 && nums[i] == nums[i-1]) continue;
            if(nums[i]>0) break;

            int lower = i+1;
            int upper = nums.size()-1;
            while(lower < upper){
                if(nums[i]+nums[lower]+nums[upper]<0)
                    lower++;
                else if(nums[i]+nums[lower]+nums[upper]>0)
                    upper--;
                else{
                    vector<int> temp({nums[i],nums[lower],nums[upper]});
                    ret.push_back(temp);
                    lower++;
                    while(lower<nums.size() && nums[i]+nums[lower]+nums[upper] == 0) lower++;
                }
            }
        }
            return ret;
    }

File: Data Structure/Array & Strings/15. 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums = sorted(nums)

        for i in range(len(nums)-2):
            if nums[i]>0:
                break
            if i == 0 or nums[i - 1] != nums[i]:
                self.sumfunc(nums, i, res)
        return res

    def sumfunc(self, nums, index, res):
        left = index+1
        right = len(nums)-1
        while left < right:
            if nums[index] + nums[left] + nums[right] == 0:
                res.append([nums[index], nums[left], nums[right]])
                left += 1
                right -= 1
                while nums[left] == nums[left - 1] and left<right:
                    left += 1
            elif nums[index] + nums[left] + nums[right] < 0:
                left += 1
            else:
                right -= 1
        return res
